<a href="https://colab.research.google.com/github/johnnyaladkhen/DM_Assignment/blob/main/CS3033_CS6405_Data_Mining_Second_Assignment_2024.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# CS3033/CS6405 - Data Mining - Second Assignment

### Submission

This assignment is **due on 14/04/24 at 23:59**. You should submit a single .ipnyb file with your python code and analysis electronically via Canvas.
Please note that this assignment will account for 25 Marks of your module grade.

### Declaration

By submitting this assignment. I agree to the following:

<font color="red">“I have read and understand the UCC academic policy on plagiarism, and agree to the requirements set out thereby in relation to plagiarism and referencing. I confirm that I have referenced and acknowledged properly all sources used in the preparation of this assignment.
I declare that this assignment is entirely my own work based on my personal study. I further declare that I have not engaged the services of another to either assist me in, or complete this assignment”</font>

### Objective

The Boolean satisfiability (SAT) problem consists in determining whether a Boolean formula is satisfiable or not. This problem is one of the most widely studied combinatorial problems in computer science. It is the classic NP-complete problem. Over the past number of decades, a significant amount of research work has focused on solving SAT problems with both complete and incomplete solvers.

An extended version of the problem is Model Counting (#SAT). In #SAT the solver needs to compute the number of solutions of a Boolean formula. A wide variety of solvers have been designed to tackle this problem.

In this project, we want to create an Algorithm Selection (AS) approach to Model Counting. For each #SAT instance, there is a specific solver that works better than the others, the goal of your machine learing approach is to classify it.

In AS we represent #SAT problems with a vector of 72 features with general information about the problem, e.g., number of variables, number of clauses, etc. There is no need to understand the features to be able to complete the assignment. For each instance, there is a 'label' column representing the name of the optimal solver.


The original dataset is available at:
https://github.com/andvise/DM_Assignment/blob/main/train_data.csv



## Data Preparation

In [1]:
import pandas as pd

df = pd.read_csv("https://github.com/andvise/DM_Assignment/blob/main/train_data.csv?raw=true")
df

c       v  clauses_vars_ratio  vars_clauses_ratio  vcg_var_mean  \
0      681.0   238.0            2.861345            0.349486      0.011143   
1      368.0   140.0            2.628571            0.380435      0.018012   
2     1935.0  1920.0            1.007812            0.992248      0.001760   
3     3452.0  2821.0            1.223680            0.817207      0.000968   
4      694.0   294.0            2.360544            0.423631      0.007656   
...      ...     ...                 ...                 ...           ...   
1331   949.0   351.0            2.703704            0.369863      0.006701   
1332  1450.0   608.0            2.384868            0.419310      0.004427   
1333   250.0   100.0            2.500000            0.400000      0.026000   
1334  4949.0  3422.0            1.446230            0.691453      0.000764   
1335   280.0   120.0            2.333333            0.428571      0.018442   

      vcg_var_coeff  vcg_var_min  vcg_var_max  vcg_var_entropy  \
0          0.905300     0.005874     0.111601         1.880038   
1          0.510753     0.005435     0.054348         1.851609   
2          1.723720     0.000000     0.012403         1.280404   
3          1.436774     0.000290     0.006083         1.192878   
4          0.493513     0.002882     0.040346         1.776102   
...             ...          ...          ...              ...   
1331       1.151888     0.002012     0.063380         1.857919   
1332       1.552830     0.002408     0.161349         1.659495   
1333       0.555766     0.000000     0.096000         1.751533   
1334       0.770643     0.000202     0.004445         1.835041   
1335       0.490373     0.000000     0.045198         1.773180   

      vcg_clause_mean  ...  gsat_FirstLocalMinStep_CoeffVariance  \
0            0.011143  ...                              0.210148   
1            0.018012  ...                              0.124438   
2            0.001760  ...                              0.066708   
3            0.000968  ...                              0.053628   
4            0.007656  ...                              0.086841   
...               ...  ...                                   ...   
1331         0.006701  ...                              0.079108   
1332         0.004427  ...                              0.062664   
1333         0.026000  ...                              0.140708   
1334         0.000764  ...                              0.032571   
1335         0.018442  ...                              0.120101   

      gsat_FirstLocalMinStep_Median  gsat_FirstLocalMinStep_Q.10  \
0                              50.0                         42.0   
1                              34.0                         29.0   
2                             102.0                         94.0   
3                             192.0                        179.0   
4                              72.0                         64.0   
...                             ...                          ...   
1331                           81.0                         73.0   
1332                          136.0                        125.0   
1333                           26.0                         21.0   
1334                          461.0                        441.0   
1335                           35.0                         30.0   

      gsat_FirstLocalMinStep_Q.90  gsat_BestAvgImprovement_Mean  \
0                            57.0                      0.954568   
1                            40.0                      1.693036   
2                           111.0                      0.398129   
3                           205.0                      0.247528   
4                            80.0                      0.822829   
...                           ...                           ...   
1331                         89.0                      0.781888   
1332                        147.0                      0.776364   
1333                         30.0           

In [2]:
# Label or target variable
df['label'].value_counts()


label
gpmc        921
d4          168
ganak       140
addmc        55
sharpsat     52
Name: count, dtype: int64

# Tasks

## Basic models and evaluation (5 Marks)

Using Scikit-learn, train and evaluate a k-NN classifier using 70% of the dataset from training and 30% for testing. For this part of the project, we are not interested in optimising the parameters; we just want to get an idea of the dataset.

In [3]:
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report, accuracy_score
from sklearn import metrics
from sklearn import model_selection
import matplotlib.pyplot as plt
# Prepare data for training as features and labels
X = df.drop('label', axis=1)
y = df['label']

# Split the dataset into 70% training and 30% testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Initializing and training the k-NN classifier
knn = KNeighborsClassifier()
knn.fit(X_train, y_train)

# Making predictions and evaluating the model
predictions = knn.predict(X_test)
accuracy = accuracy_score(y_test, predictions)
print(f'Accuracy: {accuracy}')
print(classification_report(y_test, predictions))




Accuracy: 0.6733167082294265
              precision    recall  f1-score   support

       addmc       0.50      0.69      0.58        13
          d4       0.27      0.29      0.28        49
       ganak       0.32      0.17      0.22        47
        gpmc       0.78      0.87      0.82       276
    sharpsat       0.00      0.00      0.00        16

    accuracy                           0.67       401
   macro avg       0.38      0.40      0.38       401
weighted avg       0.62      0.67      0.64       401



From the output, the overall accuracy of the classifier is approximately 67.33%.Breakdown of the performance metrics:

The model performs best for the 'gpmc' class with an f1-score of 0.82.
It performs poorly on the 'sharpsat' class
Other classes ('addmc', 'd4', 'ganak') have varying levels of performance, generally below satisfactory for a robust classifier.

## Robust evaluation (10 Marks)

In this section, we are interested in more rigorous techniques by implementing more sophisticated methods. Do your best to improve the k-NN classifier performances on this dataset.

For instance, you could consider:
* Hold-out and cross-validation.
* Hyper-parameter tuning.
* Feature selection.
* Feature normalisation.
* Etc.



Your report should provide concrete information about your reasoning, everything should be well-explained.

The key to geting good marks is to show that you evaluated different methods and that you correctly selected the configuration.

first using cross-validation.

In [4]:
from sklearn.model_selection import cross_val_score, KFold

knn = KNeighborsClassifier(n_neighbors=5)
kf = KFold(n_splits=5, shuffle=True, random_state=42)#Configures the k-fold cross-validation,The data is shuffled to ensure that the folds are randomly sampled

# Evaluating the Classifier
accuracy_scores = cross_val_score(knn, X, y, cv=kf, scoring='accuracy')
print("Accuracy scores for each fold:", accuracy_scores)
print("Mean cross-validation accuracy:", accuracy_scores.mean())


Accuracy scores for each fold: [0.67910448 0.68539326 0.71161049 0.68164794 0.70786517]
Mean cross-validation accuracy: 0.693124266308905


Accuracy: Improved to 69.31%.More stable and consistent across different data subsets, reflecting a more reliable estimate of model performance.

Trying The GridSearchCV approach

In [5]:
from sklearn.model_selection import GridSearchCV

# Setting up the parameter grid
param_grid = {'n_neighbors': range(1, 31), 'weights': ['uniform', 'distance']}

knn = KNeighborsClassifier()

# Setting up the grid search with k-fold cross-validation
grid_search = GridSearchCV(knn, param_grid, cv=5, scoring='accuracy')

# Fitting the grid search
grid_search.fit(X, y)

# Best parameters and best score
print("Best parameters:", grid_search.best_params_)
print("Best cross-validation score: {:.2f}".format(grid_search.best_score_))


Best parameters: {'n_neighbors': 9, 'weights': 'distance'}
Best cross-validation score: 0.71


Improved to 71%.The GridSearchCV approach refined the k-NN classifier by optimizing the parameters, leading to better accuracy

In [7]:
from sklearn.model_selection import train_test_split, GridSearchCV


# Normalizing features to have zero mean and unit variance
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Split the dataset into a training and a testing set
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.3, random_state=42)

# Hyperparameter tuning using GridSearchCV
param_grid = {
    'knn__n_neighbors': np.arange(1, 31),# Range of 'k' values to test
    'knn__weights': ['uniform', 'distance'],
    'knn__metric': ['euclidean', 'manhattan', 'minkowski']# Distance metrics to consider
}
pipeline = Pipeline([
    ('knn', KNeighborsClassifier())
])# Create a Pipeline with a k-NN classifier

# Setup GridSearchCV to find the best parameters for the k-NN model using the defined grid
grid_search = GridSearchCV(pipeline, param_grid, cv=5)
grid_search.fit(X_train, y_train)

# Evaluate the best model
best_model = grid_search.best_estimator_
predictions = best_model.predict(X_test)
print("Best parameters:", grid_search.best_params_)
print(classification_report(y_test, predictions))



Best parameters: {'knn__metric': 'manhattan', 'knn__n_neighbors': 10, 'knn__weights': 'distance'}
              precision    recall  f1-score   support

       addmc       0.83      0.77      0.80        13
          d4       0.50      0.33      0.40        49
       ganak       0.48      0.43      0.45        47
        gpmc       0.83      0.91      0.87       276
    sharpsat       0.36      0.25      0.30        16

    accuracy                           0.75       401
   macro avg       0.60      0.54      0.56       401
weighted avg       0.73      0.75      0.74       401



 Improved to 75.19% The latest implementation with the scaling pipeline not only achieved higher accuracy in cross-validation but also confirmed the benefits of preprocessing steps like feature scaling in k-NN classifiers

Best Parameters:
Distance Metric: manhattan - The Manhattan distance (L1 norm) was found to be the most effective
Number of Neighbors: 10 -  a slightly larger neighborhood results in better generalization than smaller ones, balancing between noise and bias effectively.
Weights: distance - This setting weights points by the inverse of their distance, giving nearer neighbors more influence on the classification.

## New classifier (10 Marks)

Replicate the previous task for a classifier different than K-NN and decision trees. Briefly describe your choice.
Try to create the best model for the given dataset.






In [12]:
# Import necessary modules
from sklearn.preprocessing import MinMaxScaler
from sklearn.decomposition import PCA
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score

# Defining a pipeline that first scales features to a range between 0 and 1, then applies PCA for dimensionality reduction  and avoid overfitting,
# and finally uses a RandomForest classifier for prediction
pipe_lr = Pipeline([
    ('scl', MinMaxScaler()),
    ('dr', PCA()),
    ('clf', RandomForestClassifier(random_state=42))  # and finally uses a RandomForest classifier for prediction
])

# Setup GridSearchCV with a specified parameter grid, focusing on 1- the number of PCA components and 2- tree depth in RandomForest
parameters = {
    "dr__n_components": [7, 9, 11, 15],
    "clf__max_depth": [3, 5, 7]
}

# Initialize GridSearchCV with the pipeline, parameter grid, and specify accuracy as the scoring metric
gs = GridSearchCV(pipe_lr, parameters, scoring="accuracy", cv=5)  # Use 5-fold cross-validation
gs.fit(X_train, y_train)  # Fit GridSearchCV on the training data

# Output the best parameters and the highest cross-validation accuracy found by GridSearchCV
print("Best parameters:", gs.best_params_)
print("Best cross-validation score:", gs.best_score_)

# Apply the best parameters found by GridSearchCV to the pipeline
pipe_lr.set_params(**gs.best_params_)
pipe_lr.fit(X_train, y_train)

# Evaluating the model’s accuracy on both the training and testing data
train_accuracy = accuracy_score(y_train, pipe_lr.predict(X_train))
test_accuracy = accuracy_score(y_test, pipe_lr.predict(X_test))

# Print the training and test set accuracies to evaluate overfitting and generalization
print("Training set accuracy:", train_accuracy)
print("Test set accuracy:", test_accuracy)


Best parameters: {'clf__max_depth': 7, 'dr__n_components': 11}
Best cross-validation score: 0.7508021390374332
Training set accuracy: 0.8673796791443851
Test set accuracy: 0.7206982543640897


The code configures a pipeline that incorporates MinMax scaling, PCA for dimensionality reduction, and a RandomForest classifier. It uses GridSearchCV to find the optimal number of PCA components and the best tree depth for the RandomForest based on accuracy.

# <font color="blue">Testing part - Not needed for CS3033</font>

Save your best model into your GitHub. And create a single code cell that loads it and evaluates it on the following test dataset:
https://github.com/andvise/DM_Assignment/blob/main/test_data.csv

We will cover this part in a Lab session.

I should be able to run the code cell independently, load all the libraries you need as well.

This link currently contains a sample of the training set. The real test set will be released after the submission.

In [15]:
from joblib import dump

# 'best_model' is obtained from GridSearchCV
dump(best_model, 'best_knn_model.joblib')

['best_knn_model.joblib']

In [13]:
from joblib import dump, load
from io import BytesIO
import requests
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report, accuracy_score
from sklearn import metrics
from sklearn import model_selection
import matplotlib.pyplot as plt
from sklearn.model_selection import cross_val_score, KFold
from sklearn.model_selection import train_test_split, GridSearchCV

# INSERT YOUR MODEL'S URL
mLink = 'url'
mfile = BytesIO(requests.get(mLink).content)
model = load(mfile)

df = pd.read_csv("https://github.com/andvise/DataAnalyticsDatasets/blob/main/test_dataset.csv")

# Your code here
# Normalizing features to have zero mean and unit variance
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Split the dataset into a training and a testing set
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.3, random_state=42)

# Hyperparameter tuning using GridSearchCV
param_grid = {
    'knn__n_neighbors': np.arange(1, 31),# Range of 'k' values to test
    'knn__weights': ['uniform', 'distance'],
    'knn__metric': ['euclidean', 'manhattan', 'minkowski']# Distance metrics to consider
}
pipeline = Pipeline([
    ('knn', KNeighborsClassifier())
])# Create a Pipeline with a k-NN classifier

# Setup GridSearchCV to find the best parameters for the k-NN model using the defined grid
grid_search = GridSearchCV(pipeline, param_grid, cv=5)
grid_search.fit(X_train, y_train)

# Evaluate the best model
best_model = grid_search.best_estimator_
predictions = best_model.predict(X_test)
print("Best parameters:", grid_search.best_params_)
print(classification_report(y_test, predictions))


MissingSchema: Invalid URL 'url': No scheme supplied. Perhaps you meant https://url?